### ISSUE TITLE
Lua: Setting visibility of fields does not work correct after hide_fields()

### ISSUE THREAD

**Issue by jeff-dh** [widelands/widelands#4746]
Hey,

I played the campaign Neptunes Revenge and couldn't finish it because I couldn't find the last artefact. After looking into the map in the editor I found it at (31,3). The point is this area is for some reason never "discovered" (e.g. the fog is  never removed) neither after I tried to discover it with an expedition nor after building a military building close to it.

I attached two screenshots. Looks for me like a bug in the -- I think you call it -- vision,  right?

I'm playing
`Widelands Version 1.0~bzr9796-202102282232 (Release)` from the daily build launchpad repository.

![neptunes-revenge-last-artefact](https://user-images.githubusercontent.com/1105041/110209582-9fb72300-7e8d-11eb-87e6-a05fed7887ec.png)
![neptunes-revenge-last-artefact-2](https://user-images.githubusercontent.com/1105041/110209585-a0e85000-7e8d-11eb-8282-bcde01c4de99.png)

**Comment by hessenfarmer** [comment]
well this is definitly a bug. thanks for reporting. The dark hexagon is the region that is shown in the beginning with the ship in danger and disclosed afterwards. It seems like the disclosure is permanent.

**Comment by frankystone** [comment]
The vision code was changed some time ago... looks like https://github.com/widelands/widelands/blob/master/data/scripting/field_animations.lua#L79 needs to be changed from:
~~~~
plr:hide_fields({region[id]},true)
~~~~
to:
~~~~
plr:hide_fields({region[id]}, false)
~~~~

Edit by hessenfarmer( sorry wanted to do a quoted reply and got the wrong menu point):
I have tried the change suggested above and the result is the fields stay visible with only light fog of war.

**Comment by frankystone** [comment]
@jeff-dh can you please upload a save game?

**Comment by jeff-dh** [comment]
[neptunes-revenge-bug.wgf.zip](https://github.com/widelands/widelands/files/6095942/neptunes-revenge-bug.wgf.zip)

I just added the .zip postfix to the .wgf file to bypass the github upload filter. It's not compressed another time, just rename it.

**Comment by hessenfarmer** [comment]
@Niektory maybe you could have a look into this as I assume something was overdone in this commit
https://github.com/widelands/widelands/commit/1878bf2ec6f146a3eb59fb862dce219916304f79#diff-930e64765522ae7ac8f73b31c02194c0f8815a0cd9ee8d714a6670e05891e472

**Comment by hessenfarmer** [comment]
the problem is that hide_fields originally was meant as set the fields to normal and unexplored. this should be the case for unexplore = false now while unexplore = true can be used to hide fields in override mode.  Currently unexplore = false does not much. it just sets the fog of war if at all.   
I tested to set the falg to false and in this situation the fields should already be partly set back to unexplored but aren't with the setting to "false".
![grafik](https://user-images.githubusercontent.com/43039887/110219455-a9597e80-7ebf-11eb-8b65-b7fbcee4c998.png)

this is the result with the settings above which is not intended
![grafik](https://user-images.githubusercontent.com/43039887/110219536-110fc980-7ec0-11eb-9661-692201060c50.png)

**Comment by Niektory** [comment]
> the problem is that hide_fields originally was meant as set the fields to normal and unexplored. this should be the case for unexplore = false now while unexplore = true can be used to hide fields in override mode. Currently unexplore = false does not much. it just sets the fog of war if at all.

The current behavior of `unexplore = false` is used in This Land is Our Land (barbarians 2) to show the tracks and then put them in the fog of war. This seems to be intended so I think this functionality should be preserved.
![shot0163](https://user-images.githubusercontent.com/4470900/111062126-5d9e6a80-84a7-11eb-8f11-6a8bcb7f1aa0.png)

Currently (after #4478) the Lua functions work as follows:
- `reveal_fields(fields)` - make fields Revealed
- `hide_fields(fields, unexplore = false)` - make fields Normal if they were Revealed
- `hide_fields(fields, unexplore = true)` - make fields Hidden

Some ways it could be resolved:

1. Probably the cleanest way would be to add a new function (not sure how to name it, `reset_fields_vision`?) that always makes fields Normal. Then it could be called after `hide_fields(fields, unexplore = true)` to achieve the desired effect in Neptune's Revenge. It would also make `hide_fields(fields, unexplore = false)` obsolete, so maybe we could get rid of the boolean argument too.

2. Alternatively add the new function and use it in This Land is Our Land instead of `hide_fields(fields, unexplore = false)`, then change the `hide_fields` behavior as in the hessenfarmer's suggestion above.

3. Alternatively change `hide_fields(fields, unexplore = false)` to always make fields Normal, then there's no need for a new function. Then we could call `hide_fields(fields, unexplore = true)` followed by `hide_fields(fields, unexplore = false)` to achieve the desired effect in Neptune's Revenge. This would probably be the simplest change, but the naming would not be ideal.

What do you think?

**Comment by frankystone** [comment]
Can't there be three states as argument?

Say `hide_fields(fields, visibility = ["hidden"|"normal"|"has_seen"]`?

* "hidden" = hide fields no matter what
* "normal" = hide fields but can be revealed again by residents, buildings and ships
* "has_seen" = hide fields like they have been seen formerly by residents, buildings and ships

For bar_02 this would be `hide_fields(fields, "has_seen")`
For Neptuns revenge this would be `hide_fields(fields, "normal")`
For situations where one wants to hide the starting field and reveal them afterwards his would be `hide_fields(fields, "hidden")`

**Comment by Noordfrees** [comment]
If I'm not mistaken the internal (C++) representation of the scripting vision magic is an "override" value that takes precedence over everything. So we should IMHO change the Lua interface to use the same mechanics. e.g.
```
set_field_vision(fields, "reveal")           -- force visible
set_field_vision(fields, "unexplored")       -- force black
set_field_vision(fields, "previously_seen")  -- force half-seen state
set_field_vision(fields, "default")          -- remove override

reveal_fields(fields)                        -- deprecated alias for set_field_vision(fields, "reveal")
hide_fields(fields, true)                    -- deprecated alias for set_field_vision(fields, "unexplored")
hide_fields(fields, false)                   -- deprecated alias for set_field_vision(fields, "default")
```

**Comment by frankystone** [comment]
If this will be named to a name which is more selfexplaining (for beginners)...

`set_field_visibility(fields, ...)`

[vision](https://www.dict.cc/?s=vision) does not fit, imho.

**Comment by hessenfarmer** [comment]
The suggestion from Nordfriese sounds good to me in combination with the suggestion of frankystone. But I think we only would need 3 values for it:
```
set_field_visibility(fields, "revealed")           -- force visible
set_field_visibility(fields, "hidden")       -- force black
set_field_visibility(fields, "normal")          -- remove override (results in previously_seen if used on revealed fields and a normal explorable field if ussed from hidden
````

**Comment by Niektory** [comment]
I'm not in the best shape right now. It might take a few more days before I get to it.

**Comment by frankystone** [comment]
Take the time you need :-)

**Comment by frankystone** [comment]
Not sure if this is related to this bug, but there is another strange behavior when placing a ship before revealing. This can be seen when changing terrains in the hidden state. Just watch this short scenario: 
[test_hide_fields.zip](https://github.com/widelands/widelands/files/6287085/test_hide_fields.zip)

In short: The problem is that changing terrain have no effect if the ship get placed before revealing. After revealing the terrains are shown like they were revealed by the ships vision range __before__ the terrains have changed although the ship get placed afterwards.

**Comment by Noordfrees** [comment]
The last one can be fixed by this one-liner change:
``` patch
diff --git a/src/logic/player.cc b/src/logic/player.cc
index 2363f2570c..2f57d6e4d1 100644
--- a/src/logic/player.cc
+++ b/src/logic/player.cc
@@ -1493,8 +1493,8 @@ void Player::hide_or_reveal_field(const Coords& coords, HideOrRevealFieldMode mo
                field.vision.set_hidden(false);
                if (!field.vision.is_visible()) {
                        field.vision = VisibleState::kVisible;
-                       rediscover_node(map, map.get_fcoords(map[i]));
                }
+               rediscover_node(map, map.get_fcoords(map[i]));
                assert(field.vision.is_visible());
                assert(!field.vision.is_revealed());
                field.vision.set_revealed(true);
```

For the initial problem, how about just letting the fields become PreviouslySeen after the intro for now, so we can apply those two quickfixes for 1.0 and postpone the real fix until after the release?
@hessenfarmer @frankystone @Niektory
